File: leetcode-python/TreeNode/TreeNode.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def __init__(self):
        self.root = TreeNode()

    def add(self, val):
        node = TreeNode(val)
        if self.root.val == 0:
            self.root = node
        else:
            myqueue = []
            treeNode = self.root
            myqueue.append(treeNode)
            while myqueue:
                treeNode = myqueue.pop(0)
                if not treeNode.left:
                    treeNode.left = node
                    return
                elif not treeNode.right:
                    treeNode.right = node
                    return
                else:
                    myqueue.append(treeNode.left)
                    myqueue.append(treeNode.right)
    # 前序遍历
    def preOrder(self, root):
        if root == None:
            return
        print(root.val)
        self.preOrder(root.left)
        self.preOrder(root.right)

    # 中序遍历
    def inOrder(self, node: TreeNode):
        if node == None:
            return
        self.inOrder(node.left)
        print(node.val)
        self.inOrder(node.right)

    # 后序遍历
    def postorder(self, root: TreeNode):
        if not root:
            return
        self.postorder(root.left)
        self.postorder(root.right)
        print(root.val)



    # 前序遍历：输出列表
    def preorderTraversal(self, root: TreeNode):
        def preOrder(root):
            if root == None:
                return
            a.append(root.val)
            preOrder(root.left)
            preOrder(root.right)
        a = []
        preOrder(root)
        return a

    def inOrderTraversal(self, root: TreeNode):
        # 实现中序遍历方法
        def inorder(root: TreeNode):
            if root == None:
                return
            inorder(root.left)
            a.append(root.val)
            inorder(root.right)
        # 创建数组队列(存储返回结点)
        a = list()
        inorder(root)
        return a

    def postorderTraversal(self, root: TreeNode):
        # 实现中序遍历方法
        def postorder(root: TreeNode):
            if not root:
                return
            postorder(root.left)
            postorder(root.right)
            a.append(root.val)

        # 创建数组队列(存储返回结点)
        a = list()
        postorder(root)
        return a

    # 层序遍历(自上而下，自左而右)
    '''
        3
       / \
      9  20
        /  \
       15   7
    [
      [3],
      [9,20],
      [15,7]
    ]
    '''
    def levelOrder(self, root: TreeNode):
        if not root: return []  # 基本情况直接返回
        queue = [root]  # 主队列
        res = []  # 初始值

        while queue:
            temp = []  # 临时队列
            res.append([i.val for i in queue])
            for item in queue:  # 把下一层的每个元素都放到队列
                if item.left: temp.append(item.left)
                if item.right: temp.append(item.right)
            queue = temp  # 把临时队列赋值给主队列
        return res






if __name__ == '__main__':
    roots = [3, 5, 7, -1, -1, 2, 6]
    tree = Solution()
    for i in roots:
        if i > 0:
            tree.add(i)
    print(tree.levelOrder(tree.root))

'''
       3
     /   \
    5     7
   / \   / \
 -1  -1 2   6

'''

File: leetcode-go/Summary/TreeNode.go
package Summary

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// 前序遍历
func preorderTraversal(root *TreeNode) (res []int) {
	var order func(*TreeNode)
	order = func(node *TreeNode) {
		if node == nil {
			return
		}
		res = append(res, node.Val) // 前序
		order(node.Left)
		order(node.Right)
	}
	if root == nil {
		return nil
	}
	order(root)
	return res
}

// 中序遍历
func inorderTraversal(root *TreeNode) (res []int) {
	var order func(*TreeNode)
	order = func(node *TreeNode) {
		if node == nil {
			return
		}
		order(node.Left)
		res = append(res, node.Val) // 中序
		order(node.Right)
	}
	if root == nil {
		return nil
	}
	order(root)
	return
}

// 后序遍历
func postorderTraversal(root *TreeNode) (res []int) {
	var order func(*TreeNode)
	order = func(node *TreeNode) {
		if node == nil {
			return
		}
		order(node.Left)
		order(node.Right)
		res = append(res, node.Val) // 后序
	}
	if root == nil {
		return nil
	}
	order(root)
	return
}

// 层序遍历
func levelOrder(root *TreeNode) (res [][]int) {
	var dfs func(node *TreeNode, level int)
	dfs = func(node *TreeNode, level int) {
		if len(res) == level {
			res = append(res, []int{})
		}
		res[level] = append(res[level], node.Val)
		if node.Left != nil {
			dfs(node.Left, level+1)
		}
		if node.Right != nil {
			dfs(node.Right, level+1)
		}

	}

	if root == nil {
		return res
	}
	dfs(root, 0)
	return res
}

// 二叉树的最大深度
func maxDepth(root *TreeNode) int {
	if root == nil {
		return 0
	}
	return max(maxDepth(root.Left), maxDepth(root.Right)) + 1
}

// 二叉树的最小深度
func minDepth(root *TreeNode) int {
	if root == nil {
		return 0
	}
	m1 := minDepth(root.Left)
	m2 := minDepth(root.Right)
	if root.Left == nil || root.Right == nil {
		return m1 + m2 + 1
	}
	return min(m1, m2) + 1
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func min(a, b int) int {
	if a > b {
		return b
	}
	return a
}

// 对称二叉树
func isSymmetric(root *TreeNode) bool {
	if root == nil {
		return true
	}

	return check(root.Left, root.Right)
}

func check(root1 *TreeNode, root2 *TreeNode) bool {
	if root1 == nil && root2 == nil {
		return true
	}
	if root1 == nil || root2 == nil {
		return false
	}

	return root1.Val == root2.Val && check(root1.Left, root2.Right) && check(root1.Right, root2.Left)
}
